README.md
pyads - Python package
======================

[![PyPI version](https://badge.fury.io/py/pyads.svg)](https://badge.fury.io/py/pyads)
[![Anaconda-Server Badge](https://anaconda.org/conda-forge/pyads/badges/version.svg)](https://anaconda.org/conda-forge/pyads)
[![Anaconda-Server Badge](https://anaconda.org/conda-forge/pyads/badges/platforms.svg)](https://anaconda.org/conda-forge/pyads)

[![CI](https://github.com/stlehmann/pyads/actions/workflows/ci.yml/badge.svg)](https://github.com/stlehmann/pyads/actions/workflows/ci.yml)
[![Coverage Status](https://coveralls.io/repos/github/stlehmann/pyads/badge.svg?branch=master)](https://coveralls.io/github/stlehmann/pyads?branch=master)
[![Documentation Status](https://readthedocs.org/projects/pyads/badge/?version=latest)](http://pyads.readthedocs.io/en/latest/?badge=latest)
[![Downloads](https://pepy.tech/badge/pyads)](https://pepy.tech/project/pyads)
[![Downloads](https://pepy.tech/badge/pyads/week)](https://pepy.tech/project/pyads)

This is a python wrapper for TwinCATs ADS library. It provides python functions
for communicating with TwinCAT devices. *pyads* uses the C API provided by *TcAdsDll.dll* on Windows *adslib.so* on Linux. The documentation for the ADS API is available on [infosys.beckhoff.com](http://infosys.beckhoff.com/english.php?content=../content/1033/tcadsdll2/html/tcadsdll_api_overview.htm&id=20557).

Documentation: http://pyads.readthedocs.io/en/latest/index.html

# Installation

From PyPi:

```bash
pip install pyads
```

From conda-forge:

```bash
conda install pyads
```

From source:

```bash
git clone https://github.com/MrLeeh/pyads.git --recursive
cd pyads
python setup.py install
```

## Features

* connect to a remote TwinCAT device like a plc or a PC with TwinCAT
* create routes on Linux devices and on remote plcs
* supports TwinCAT 2 and TwinCAT 3
* read and write values by name or address
* read and write DUTs (structures) from the plc
* notification callbacks

## Basic usage

```python
import pyads

# connect to plc and open connection
target_netid = '5.12.82.20.1.1' # The Ams Net Id of the target device. In TwinCAT3 created from device MAC address by default 
plc = pyads.Connection(target_netid, pyads.PORT_TC3PLC1)
plc.open()

# read int value by name
i = plc.read_by_name("GVL.int_val")

# write int value by name
plc.write_by_name("GVL.int_val", i)

# close connection
plc.close()
```

[0]: https://infosys.beckhoff.de/english.php?content=../content/1033/TcSystemManager/Basics/TcSysMgr_AddRouteDialog.htm&id=

## Contributing guidelines

Contributions are very much welcome. pyads is under active development. However it is a side-project of mine so please have some
patience when creating issues or PRs. Here are some main guidelines which I ask you to follow along:

* Create PRs based on the [master](https://github.com/stlehmann/pyads) branch.
* Add an entry to the [Changelog](https://github.com/stlehmann/pyads/blob/master/CHANGELOG.md).
* Keep PRs small (if possible), this makes reviews easier and your PR can be merged faster.
* Address only one issue per PR. If you want to make additional fixes e.g. on import statements, style or documentation 
which are not directly related to your issue please create an additional PR that adresses these small fixes.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

## 3.3.5 [unreleased]

### Added
* Added `PORT_TC3PLC1`, `PORT_TC3PLC2`, `PORT_TC3PLC3` and `PORT_TC3PLC4` to constants which are used by TwinCAT3 PLC programs
* [#223](https://github.com/stlehmann/pyads/pull/223) Add structure support for symbols
* [#238](https://github.com/stlehmann/pyads/pull/238) Add LINT type to DATATYPE_MAP

### Changed
* Changed the default connection port to `PORT_TC3PLC1` in the README.md as this is the default for TwinCAT3
* [#221](https://github.com/stlehmann/pyads/pull/221) CI now uses Github Actions instead of TravisCI. Also Upload to PyPi is now on automatic.
* [#242](https://github.com/stlehmann/pyads/pull/242) Upgrade requirements.txt
* Use TwinCAT3 default port 851 (PORT_TC3PLC1) in docs

### Removed

## 3.3.4

### Added
* [#187](https://github.com/stlehmann/pyads/pull/187) Support structured data types in `read_list_by_name`
* [#220](https://github.com/stlehmann/pyads/pull/220) Support structured data types in `write_list_by_name`. Also the
  AdvancedHandler of the testserver now support sumup_read and sumup_write commands.
* [#195](https://github.com/stlehmann/pyads/pull/195) Read/write by name without passing the datatype
* [#200](https://github.com/stlehmann/pyads/pull/200) Split read write by list into max-ads-sub-comands chunks
* [#206](https://github.com/stlehmann/pyads/pull/206) AdsSymbol now supports DT, DATE_TIME and TIME datatypes 

### Changed
* [#202](https://github.com/stlehmann/pyads/pull/202) Testserver now support variable sumread and sumwrite with 
  variable length for uint8 and string datatypes
* [#209](https://github.com/stlehmann/pyads/pull/209) Removed duplicate tests and added addtional asserts to existing tests
* [#212](https://github.com/stlehmann/pyads/pull/212) Add type annotations to structs.py

### Removed

## 3.3.3

### Added
* comprehensive documentation and short Quickstart guide

### Changed
* [#192](https://github.com/stlehmann/pyads/pull/192) Make AdsSymbol even more pythonic
  * Replace AdsSymbol.set_auto_update function by AdsSymbol.auto_update property
  * Make AdsSymbol.value a property
  * AdsSymbol.value setter writes to plc if AdsSymbol.auto_update is True

### Removed
* [#193](https://github.com/stlehmann/pyads/pull/193) Remove testserver_ex package which is still in development. 
  The testserver_ex package can be found in the [testserver_ex branch](https://github.
  com/stlehmann/pyads/tree/testserver_ex).

## 3.3.2

### Added

### Changed
* fixed error with source distribution not containing adslib directory
 
### Removed

## 3.3.1

### Added
* [#174](https://github.com/stlehmann/pyads/pull/174) Add `AdsSymbol` class for pythonic access
* [#169](https://github.com/stlehmann/pyads/pull/169) Add adsGetNetIdForPLC to pyads_ex
* [#179](https://github.com/stlehmann/pyads/pull/179) Added destructor to `pyads.Connection`

### Changed

### Removed

## 3.3.0

### Added
* [#155](https://github.com/stlehmann/pyads/pull/155) Add get_all_symbols method to Connection
* [#157](https://github.com/stlehmann/pyads/pull/157) Add write_structure_by_name method to Connection
* [#161](https://github.com/stlehmann/pyads/pull/161) Add sum read and write commands

### Changed
* [#150](https://github.com/stlehmann/pyads/pull/150) Use function annotations and variable annotations for type annotations

### Removed
* [#152](https://github.com/stlehmann/pyads/pull/152) Remove deprecated functions
* [#150](https://github.com/stlehmann/pyads/pull/150) Drop support for Python 2.7 and Python 3.5

## 3.2.2

### Added
* [#141](https://github.com/stlehmann/pyads/issues/141) Add ULINT support to read_structure_by_name
* [#143](https://github.com/stlehmann/pyads/issues/143) Add parse_notification method to Connection

### Changed
* [#140](https://github.com/stlehmann/pyads/pull/140) Fix lineendings to LF in the repository
* [#139](https://github.com/stlehmann/pyads/pull/139) Fix documentation and test issues with DeviceNotifications
* [ea707](https://github.com/stlehmann/pyads/tree/ea7073d93feac75c1864d1fe8ab2e14a2068b552) Fix documentation on
 ReadTheDocs
* [45859](https://github.com/stlehmann/pyads/tree/45859d6e9038b55d319efdbda95d3d6eeadd45e3) Fix issue with async handling in adslib

### Removed

## 3.2.1

### Added
* [#130](https://github.com/stlehmann/pyads/pull/130) Allow read_write with NULL read/write data
* [#131](https://github.com/stlehmann/pyads/pull/131) Add FILETIME passthrough to notification decorator

### Changed
* [#135](https://github.com/stlehmann/pyads/pull/135) Bug fix for setting up remote route from linux
* [#137](https://github.com/stlehmann/pyads/pull/137) Update adslib from upstream

### Removed

## 3.2.0

### Added
* [#111](https://github.com/stlehmann/pyads/pull/111) test cases for notification decorators
* [#113](https://github.com/stlehmann/pyads/pull/113) Add option not to check for data size
* [#118](https://github.com/stlehmann/pyads/pull/118) Add support for arrays in notification decorator
* [#112](https://github.com/stlehmann/pyads/pull/112) Add getters/setters for connection netid and port 

### Changed
* [#128](https://github.com/stlehmann/pyads/pull/128) Deprecation warning for older non-class functions. In
future versions only methods of the Connection class are supported.

### Removed
* [#127](https://github.com/stlehmann/pyads/pull/127) Drop support for Python 2

## 3.1.3

### Added
* [#120](https://github.com/stlehmann/pyads/pull/120) Allow to write ctypes directly

### Changed
* [#125](https://github.com/stlehmann/pyads/pull/125) Add notifications by address. The `data_name
` parameter changed to `data` as now not only strings can be passed but also a tuple with index group and offset.
* [#123](https://github.com/stlehmann/pyads/pull/123) Add ULINT data type

pyads/__init__.py
"""The pyads package.

:author: Stefan Lehmann <stlm@posteo.de>
:license: MIT, see license file or https://opensource.org/licenses/MIT
:created on: 2018-06-11 18:15:53

"""
from .structs import AmsAddr, NotificationAttrib

from .ads import (
    open_port,
    close_port,
    get_local_address,
    add_route,
    add_route_to_plc,
    delete_route,
    Connection,
    set_local_address,
    set_timeout,
    size_of_structure,
    dict_from_bytes,
    bytes_from_dict,
    StructureDef,
)

from .pyads_ex import ADSError

from .constants import (
    PLCTYPE_BOOL,
    PLCTYPE_BYTE,
    PLCTYPE_DATE,
    PLCTYPE_DINT,
    PLCTYPE_DT,
    PLCTYPE_DWORD,
    PLCTYPE_INT,
    PLCTYPE_LINT,
    PLCTYPE_LREAL,
    PLCTYPE_REAL,
    PLCTYPE_SINT,
    PLCTYPE_STRING,
    PLCTYPE_TIME,
    PLCTYPE_TOD,
    PLCTYPE_ULINT,
    PLCTYPE_UDINT,
    PLCTYPE_UINT,
    PLCTYPE_USINT,
    PLCTYPE_WORD,
    PLCTYPE_ARR_DINT,
    PLCTYPE_ARR_INT,
    PLCTYPE_ARR_LREAL,
    PLCTYPE_ARR_REAL,
    PLCTYPE_ARR_SHORT,
    PLC_DEFAULT_STRING_SIZE,
    DATATYPE_MAP,
)

from .constants import (
    PORT_EVENTLOGGER,
    PORT_IO,
    PORT_LOGGER,
    PORT_NC,
    PORT_NOCKE,
    PORT_SCOPE,
    PORT_SPECIALTASK1,
    PORT_SPECIALTASK2,
    PORT_SPS1,
    PORT_SPS2,
    PORT_SPS3,
    PORT_SPS4,
    PORT_SYSTEMSERVICE,
    PORT_CAM,
    PORT_TC2PLC1,
    PORT_TC2PLC2,
    PORT_TC2PLC3,
    PORT_TC2PLC4,
    PORT_TC3PLC1,
    PORT_TC3PLC2,
    PORT_TC3PLC3,
    PORT_TC3PLC4,
)

from .constants import (
    INDEXGROUP_MEMORYBYTE,
    INDEXGROUP_MEMORYBIT,
    INDEXGROUP_MEMORYSIZE,
    INDEXGROUP_RETAIN,
    INDEXGROUP_RETAINSIZE,
    INDEXGROUP_DATA,
    INDEXGROUP_DATASIZE,
)

from .constants import (
    ADSSTATE_INVALID,
    ADSSTATE_IDLE,
    ADSSTATE_RESET,
    ADSSTATE_INIT,
    ADSSTATE_START,
    ADSSTATE_RUN,
    ADSSTATE_STOP,
    ADSSTATE_SAVECFG,
    ADSSTATE_LOADCFG,
    ADSSTATE_POWERFAILURE,
    ADSSTATE_POWERGOOD,
    ADSSTATE_ERROR,
    ADSSTATE_SHUTDOWN,
    ADSSTATE_SUSPEND,
    ADSSTATE_RESUME,
    ADSSTATE_CONFIG,
    ADSSTATE_RECONFIG,
)

from .constants import (
    ADSTRANS_NOTRANS,
    ADSTRANS_CLIENTCYCLE,
    ADSTRANS_CLIENT1REQ,
    ADSTRANS_SERVERCYCLE,
    ADSTRANS_SERVERONCHA,
)

from .symbol import AdsSymbol

__version__ = '3.3.4'


pyads/constants.py
# -*- coding: utf-8 -*-
"""Constants for the work with the ADS API.

:author: Stefan Lehmann <stlm@posteo.de>
:license: MIT, see license file or https://opensource.org/licenses/MIT
:created on: 2018-06-11 18:15:53

"""
from typing import Type, Dict, Callable, Union
from ctypes import (
    Array,
    c_bool,
    c_ubyte,
    c_int8,
    c_uint8,
    c_int16,
    c_uint16,
    c_int32,
    c_uint32,
    c_float,
    c_double,
    c_char,
    c_int64,
    c_uint64,
)

STRING_BUFFER: int = 1024
PLC_DEFAULT_STRING_SIZE: int = 80

MAX_ADS_SUB_COMMANDS: int = 500

# plc data types:
PLCTYPE_BOOL = c_bool
PLCTYPE_BYTE = c_ubyte
PLCTYPE_DWORD = c_uint32
PLCTYPE_DINT = c_int32
PLCTYPE_INT = c_int16
PLCTYPE_LREAL = c_double
PLCTYPE_REAL = c_float
PLCTYPE_SINT = c_int8
PLCTYPE_STRING = c_char
PLCTYPE_TOD = c_int32
PLCTYPE_UBYTE = c_ubyte
PLCTYPE_UDINT = c_uint32
PLCTYPE_UINT = c_uint16
PLCTYPE_USINT = c_uint8
PLCTYPE_WORD = c_uint16
PLCTYPE_LINT = c_int64
PLCTYPE_ULINT = c_uint64

PLCTYPE_DATE = PLCTYPE_DWORD
PLCTYPE_DATE_AND_TIME = PLCTYPE_DWORD
PLCTYPE_DT = PLCTYPE_DWORD
PLCTYPE_TIME = PLCTYPE_DWORD

# Typing
PLCSimpleDataType = Union[
    PLCTYPE_BOOL,
    PLCTYPE_BYTE,
    PLCTYPE_DWORD,
    PLCTYPE_DINT,
    PLCTYPE_INT,
    PLCTYPE_LREAL,
    PLCTYPE_REAL,
    PLCTYPE_SINT,
    PLCTYPE_STRING,
    PLCTYPE_TOD,
    PLCTYPE_UBYTE,
    PLCTYPE_UDINT,
    PLCTYPE_UINT,
    PLCTYPE_USINT,
    PLCTYPE_WORD,
    PLCTYPE_LINT,
    PLCTYPE_ULINT,
    PLCTYPE_DATE,
    PLCTYPE_DATE_AND_TIME,
    PLCTYPE_DT,
    PLCTYPE_TIME,
]
PLCDataType = Union[Array, PLCSimpleDataType]

# Datatype unpacking values
DATATYPE_MAP: Dict[Type, str] = {
    PLCTYPE_BOOL: "<?",
    PLCTYPE_BYTE: "<B",
    PLCTYPE_DINT: "<i",
    PLCTYPE_DWORD: "<I",
    PLCTYPE_INT: "<h",
    PLCTYPE_LREAL: "<d",
    PLCTYPE_REAL: "<f",
    PLCTYPE_SINT: "<b",
    PLCTYPE_UDINT: "<I",
    PLCTYPE_UINT: "<H",
    PLCTYPE_USINT: "<B",
    PLCTYPE_ULINT: "<Q",
    PLCTYPE_WORD: "<H",
}


# ADS data types
ADST_VOID: int = 0
ADST_INT8: int = 16
ADST_UINT8: int = 17
ADST_INT16: int = 2
ADST_UINT16: int = 18
ADST_INT32: int = 3
ADST_UINT32: int = 19
ADST_INT64: int = 20
ADST_UINT64: int = 21
ADST_REAL32: int = 4
ADST_REAL64: int = 5
ADST_BIGTYPE: int= 65
ADST_STRING: int = 30
ADST_WSTRING: int = 31
ADST_REAL80: int = 32
ADST_BIT: int = 33
ADST_MAXTYPES: int = 34


ads_type_to_ctype = {
    # ADST_VOID
    ADST_INT8: PLCTYPE_BYTE,
    ADST_UINT8: PLCTYPE_UBYTE,
    ADST_INT16: PLCTYPE_INT,
    ADST_UINT16: PLCTYPE_UINT,
    ADST_INT32: PLCTYPE_DINT,
    ADST_UINT32: PLCTYPE_UDINT,
    ADST_INT64: PLCTYPE_LINT,
    ADST_UINT64: PLCTYPE_ULINT,
    ADST_REAL32: PLCTYPE_REAL,
    ADST_REAL64: PLCTYPE_LREAL,
    # ADST_BIGTYPE
    ADST_STRING: PLCTYPE_STRING,
    # ADST_WSTRING
    # ADST_REAL80
    ADST_BIT: PLCTYPE_BOOL,
}


def PLCTYPE_ARR_REAL(n: int) -> Type[Array]:
    """Return an array with n float values."""
    return c_float * n


def PLCTYPE_ARR_LREAL(n: int) -> Type[Array]:
    """Return an array with n double values."""
    return c_double * n


def PLCTYPE_ARR_BOOL(n: int) -> Type[Array]:
    """Return an array with n boolean values."""
    return c_bool * n


def PLCTYPE_ARR_INT(n: int) -> Type[Array]:
    """Return an array with n int16 values."""
    return c_int16 * n


def PLCTYPE_ARR_UINT(n: int) -> Type[Array]:
    """Return an array with n uint16 values."""
    return c_uint16 * n


def PLCTYPE_ARR_SHORT(n: int) -> Type[Array]:
    """Return an array with n short values."""
    return c_int16 * n


def PLCTYPE_ARR_USHORT(n: int) -> Type[Array]:
    """Return an array with n ushort values."""
    return c_uint16 * n


def PLCTYPE_ARR_DINT(n: int) -> Type[Array]:
    """Return an array with n int32 values."""
    return c_int32 * n


def PLCTYPE_ARR_UDINT(n: int) -> Type[Array]:
    """Return an array with n uint32 values."""
    return c_uint32 * n


def PLCTYPE_ARR_SINT(n: int) -> Type[Array]:
    """Return an array with n int8 values."""
    return c_int8 * n


def PLCTYPE_ARR_USINT(n: int) -> Type[Array]:
    """Return an array with n uint8 values."""
    return c_uint8 * n


# Map c-type array names to PLCTYPE_* arrays
PLC_ARRAY_MAP: Dict[str, Callable] = {
    'real': PLCTYPE_ARR_LREAL,  # LREAL, not REAL
    'boolean': PLCTYPE_ARR_BOOL,
    'int32': PLCTYPE_ARR_DINT,
    'uint32': PLCTYPE_ARR_UDINT,
    'int16': PLCTYPE_ARR_INT,
    'uint16': PLCTYPE_ARR_UINT,
    'int8': PLCTYPE_ARR_SINT,
    'uint8': PLCTYPE_ARR_USINT,
}


# Index Group
# READ_M - WRITE_M
INDEXGROUP_MEMORYBYTE = 0x4020  #: plc memory area (%M), offset means byte-offset
# READ_MX - WRITE_MX
INDEXGROUP_MEMORYBIT = (
    0x4021
)  #: plc memory area (%MX), offset means the bit address, calculatedb by bytenumber * 8 + bitnumber  # noqa: E501
# PLCADS_IGR_RMSIZE
INDEXGROUP_MEMORYSIZE = 0x4025  #: size of the memory area in bytes
# PLCADS_IGR_RWRB
INDEXGROUP_RETAIN = 0x4030  #: plc retain memory area, offset means byte-offset
# PLCADS_IGR_RRSIZE
INDEXGROUP_RETAINSIZE = 0x4035  #: size of the retain area in bytes
# PLCADS_IGR_RWDB
INDEXGROUP_DATA = 0x4040  #: data area, offset means byte-offset
# PLCADS_IGR_RDSIZE
INDEXGROUP_DATASIZE = 0x4045  #: size of the data area in bytes


ADSIGRP_SYMTAB = 0xF000
ADSIGRP_SYMNAME = 0xF001
ADSIGRP_SYMVAL = 0xF002

ADSIGRP_SYM_HNDBYNAME = 0xF003
ADSIGRP_SYM_VALBYNAME = 0xF004
ADSIGRP_SYM_VALBYHND = 0xF005
ADSIGRP_SYM_RELEASEHND = 0xF006
ADSIGRP_SYM_INFOBYNAME = 0xF007
ADSIGRP_SYM_VERSION = 0xF008
ADSIGRP_SYM_INFOBYNAMEEX = 0xF009

ADSIGRP_SYM_DOWNLOAD = 0xF00A
ADSIGRP_SYM_UPLOAD = 0xF00B
ADSIGRP_SYM_UPLOADINFO = 0xF00C
ADSIGRP_SYM_DOWNLOAD2 = 0xF00D
ADSIGRP_SYM_DT_UPLOAD = 0xF00E
ADSIGRP_SYM_UPLOADINFO2 = 0xF00F

ADSIGRP_SYMNOTE = 0xF010  #: notification of named handle
ADSIGRP_IOIMAGE_RWIB = 0xF020  #: read/write input byte(s)
ADSIGRP_IOIMAGE_RWIX = 0xF021  #: read/write input bit
ADSIGRP_IOIMAGE_RWOB = 0xF030  #: read/write output byte(s)
ADSIGRP_IOIMAGE_RWOX = 0xF031  #: read/write output bit
ADSIGRP_IOIMAGE_CLEARI = 0xF040  #: write inputs to null
ADSIGRP_IOIMAGE_CLEARO = 0xF050  #: write outputs to null

ADSIGRP_SUMUP_READ = 0xF080  #: ADS Sum Read Request
ADSIGRP_SUMUP_WRITE = 0xF081  #: ADS Sum Write Request

ADSIGRP_DEVICE_DATA = 0xF100  #: state, name, etc...
ADSIOFFS_DEVDATA_ADSSTATE = 0x0000  #: ads state of device
ADSIOFFS_DEVDATA_DEVSTATE = 0x0002  #: device state


# PORTS
PORT_LOGGER: int = 100
PORT_EVENTLOGGER: int = 110
PORT_IO: int = 300
PORT_SPECIALTASK1: int = 301
PORT_SPECIALTASK2: int = 302
PORT_NC: int = 500
PORT_SPS1: int = 801
PORT_SPS2: int = 811
PORT_SPS3: int = 821
PORT_SPS4: int = 831
PORT_TC2PLC1: int = PORT_SPS1  #: TwinCAT2 PLC Runtime System 1
PORT_TC2PLC2: int = PORT_SPS2  #: TwinCAT2 PLC Runtime System 2
PORT_TC2PLC3: int = PORT_SPS3  #: TwinCAT2 PLC Runtime System 3
PORT_TC2PLC4: int = PORT_SPS4  #: TwinCAT2 PLC Runtime System 4
PORT_TC3PLC1: int = 851  #: TwinCAT3 PLC Runtime System 1
PORT_TC3PLC2: int = 852  #: TwinCAT3 PLC Runtime System 2
PORT_TC3PLC3: int = 853  #: TwinCAT3 PLC Runtime System 3
PORT_TC3PLC4: int = 854  #: TwinCAT3 PLC Runtime System 4
PORT_NOCKE: int = 900
PORT_CAM: int = PORT_NOCKE
PORT_SYSTEMSERVICE: int = 10000
PORT_SCOPE: int = 14000
PORT_REMOTE_UDP: int = 48899

# ADSState-constants
ADSSTATE_INVALID: int = 0
ADSSTATE_IDLE: int = 1
ADSSTATE_RESET: int = 2
ADSSTATE_INIT: int = 3
ADSSTATE_START: int = 4
ADSSTATE_RUN: int = 5
ADSSTATE_STOP: int = 6
ADSSTATE_SAVECFG: int = 7
ADSSTATE_LOADCFG: int = 8
ADSSTATE_POWERFAILURE: int = 9
ADSSTATE_POWERGOOD: int = 10
ADSSTATE_ERROR: int = 11
ADSSTATE_SHUTDOWN: int = 12
ADSSTATE_SUSPEND: int = 13
ADSSTATE_RESUME: int = 14
ADSSTATE_CONFIG: int = 15
ADSSTATE_RECONFIG: int = 16

# ADSTransmode
ADSTRANS_NOTRANS: int = 0  #: no notifications
ADSTRANS_CLIENTCYCLE: int = 1
ADSTRANS_CLIENT1REQ: int = 2
ADSTRANS_SERVERCYCLE: int = 3  #: notify on a cyclic base
ADSTRANS_SERVERONCHA: int = 4  #: notify everytime the value changes

# symbol flags
ADSSYMBOLFLAG_PERSISTENT = 0x00000001
ADSSYMBOLFLAG_BITVALUE = 0x00000002
ADSSYMBOLFLAG_REFERENCETO = 0x0004
ADSSYMBOLFLAG_TYPEGUID = 0x0008
ADSSYMBOLFLAG_TCCOMIFACEPTR = 0x0010
ADSSYMBOLFLAG_READONLY = 0x0020
ADSSYMBOLFLAG_CONTEXTMASK = 0x0F00

# ADS Command IDs
ADSCOMMAND_INVALID = 0x00
ADSCOMMAND_READDEVICEINFO = 0x01
ADSCOMMAND_READ = 0x02
ADSCOMMAND_WRITE = 0x03
ADSCOMMAND_READSTATE = 0x04
ADSCOMMAND_WRITECTRL = 0x05
ADSCOMMAND_ADDDEVICENOTE = 0x06
ADSCOMMAND_DELDEVICENOTE = 0x07
ADSCOMMAND_DEVICENOTE = 0x08
ADSCOMMAND_READWRITE = 0x09

# STATE Flags
ADSSTATEFLAG_REQRESP = 0x0001
ADSSTATEFLAG_COMMAND = 0x0004
